crates/p2p/src/swarm/message.rs

```diff
@@ -0,0 +1,148 @@
+//! Message types for P2P protocol communication.
+
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use libp2p::{PeerId, identity::PublicKey};
+use serde::{Deserialize, Serialize};
+
+use super::errors::{SetupError, SetupUpgradeError};
+
+pub(super) mod opaque_serializer {
+    use serde::{self, Deserializer, Serializer, de};
+
+    use super::PublicKey;
+
+    pub(super) fn serialize<S>(data: &PublicKey, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_bytes(&data.encode_protobuf())
+    }
+
+    pub(super) fn deserialize<'de, D>(deserializer: D) -> Result<PublicKey, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let bytes: Vec<u8> = serde::Deserialize::deserialize(deserializer)?;
+        PublicKey::try_decode_protobuf(&bytes).map_err(de::Error::custom)
+    }
+}
+
+/// Protocol version for all messages.
+pub(crate) const PROTOCOL_VERSION: u8 = 2;
+
+/// Protocol identifier for setup messages.
+pub(crate) const SETUP_PROTOCOL_ID: u8 = 1;
+
+/// Wrapper for signed messages.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SignedMessage {
+    /// The serialized message content.
+    pub message: Vec<u8>,
+    /// The signature of the message.
+    pub signature: Vec<u8>,
```

Comment:
Hey @Arniiiii can this `signature` not be a `Vec<u8>`? We know the size of the signature because it is an `ed25519` signature.
Despite `libp2p` making it a `Vec<u8>` (https://docs.rs/libp2p-identity/0.2.11/src/libp2p_identity/ed25519.rs.html#65-67) we know that it is 64 bytes because that's what `ed25519_dalek` spits out and what `libp2p` uses under the hood (https://docs.rs/ed25519-dalek/latest/ed25519_dalek/struct.Signature.html#method.to_bytes)

Reply:
We can at least make `[u8; 64]` with some effort.